File: src/gameRules.ts
/**
 * Mirror of plutus/GameRules.hs
 *
 * PrizeTable defines base multipliers for each winning tier.
 * Payout = baseForTier(tier) * priceUsdm / 2
 *
 * USDM SUB-UNITS: 1 USDM = 100 integer units.
 * pdPriceUsdm is stored in sub-units (e.g. Genesis = 100, Class 1 = 200).
 * The payout formula is: payout = base * priceUsdm / 2
 * With priceUsdm in sub-units, the result is in sub-units:
 *   payout_subunits = base * priceUsdm / 2
 *   payout_USDM = payout_subunits / 100
 *
 * Example (Genesis, priceUsdm = 100):
 *   Tier 1: 2 * 100 / 2 = 100 sub-units = 1.00 USDM
 *   Tier 2: 5 * 100 / 2 = 250 sub-units = 2.50 USDM
 *   Tier 3: 10 * 100 / 2 = 500 sub-units = 5.00 USDM
 *   Tier 4: 200 * 100 / 2 = 10000 sub-units = 100.00 USDM
 *   Tier 5: 1000 * 100 / 2 = 50000 sub-units = 500.00 USDM
 *
 * Ticket CLASS (pdPriceUsdm) != winning TIER.
 * A 1 USDM ticket can win tier 5 (500 USDM).
 * A 100 USDM ticket can win tier 1 (100 USDM).
 */

import { sha256, concatBytes } from './beacon'

export type PrizeTable = {
  tier1: number
  tier2: number
  tier3: number
  tier4: number
  tier5: number
}

export const defaultPrizeTable: PrizeTable = {
  tier1: 2,
  tier2: 5,
  tier3: 10,
  tier4: 200,
  tier5: 1000,
}

function baseForTier(t: PrizeTable, tier: number): number {
  if (tier === 1) return t.tier1
  if (tier === 2) return t.tier2
  if (tier === 3) return t.tier3
  if (tier === 4) return t.tier4
  if (tier === 5) return t.tier5
  return 0
}

/**
 * Returns payout in USDM sub-units (1 USDM = 100 sub-units).
 *
 * Parity with Plutus GameRules.prizeAmountForTier:
 *   Plutus: (baseForTier tier * priceUsdm) `divide` 2
 *   TypeScript: Math.floor((baseForTier tier * priceUsdm) / 2)
 *
 *   Both produce identical results for positive integers.
 *   The divisor is 2 in both. Sub-unit representation comes from
 *   priceUsdm being stored in sub-units (e.g. 100 for Genesis = 1 USDM).
 *
 *   Example (Genesis, priceUsdm = 100):
 *     Tier 1: 2 * 100 / 2 = 100 sub-units = 1.00 USDM
 *     Tier 2: 5 * 100 / 2 = 250 sub-units = 2.50 USDM
 *     Tier 3: 10 * 100 / 2 = 500 sub-units = 5.00 USDM
 *     Tier 4: 200 * 100 / 2 = 10000 sub-units = 100.00 USDM
 *     Tier 5: 1000 * 100 / 2 = 50000 sub-units = 500.00 USDM
 */
export function prizeAmountForTier(
  table: PrizeTable,
  tier: number,
  priceUsdm: number
): number {
  if (tier <= 0 || priceUsdm <= 0) return 0
  return Math.floor((baseForTier(table, tier) * priceUsdm) / 2)
}

export function classifyTier(symbols: Uint8Array): number {
  if (symbols.length < 6) return 0
  for (let sym = 5; sym >= 1; sym--) {
    let count = 0
    for (let i = 0; i < 6; i++) {
      if (symbols[i] === sym) count++
    }
    if (count >= 3) return sym
  }
  return 0
}

/** Must match GameRules.generateSymbols on-chain (rejection sampling, bounded within 32 bytes). */
export async function generateSymbols(symbolsSeed: Uint8Array): Promise<Uint8Array> {
  const out = new Uint8Array(6)
  let count = 0
  let hashPos = 0
  while (count < 6) {
    if (hashPos >= 32) {
      throw new Error('generateSymbols: hash exhausted for position')
    }
    const h = await sha256(concatBytes(new Uint8Array([count]), symbolsSeed))
    const byte = h[hashPos]
    if (byte === 255) {
      hashPos++
    } else {
      out[count] = (byte % 5) + 1
      count++
      hashPos = 0
    }
  }
  return out
}

export async function resultBinding(
  digest: Uint8Array,
  symbols: Uint8Array,
  fieldFn: (bs: Uint8Array) => Uint8Array,
  sha: (d: Uint8Array) => Promise<Uint8Array>
): Promise<Uint8Array> {
  return sha(concatBytes(fieldFn(digest), fieldFn(symbols)))
}

File: plutus/GameRules.hs
{-# LANGUAGE DataKinds             #-}
{-# LANGUAGE FlexibleInstances     #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE NoImplicitPrelude     #-}
{-# LANGUAGE ScopedTypeVariables   #-}
{-# LANGUAGE TemplateHaskell       #-}
{-# LANGUAGE UndecidableInstances  #-}
{-# LANGUAGE ViewPatterns          #-}

module GameRules
  ( PrizeTable (..)
  , classifyTier
  , prizeAmountForTier
  , defaultPrizeTable
  , generateSymbols
  ) where

import PlutusTx
import PlutusTx.Prelude

-- ============================================================
-- Prize table
-- ============================================================

-- | PrizeTable defines base multipliers for each winning tier.
--
-- The payout for a ticket is: baseForTier(tier) * priceUsdm / 2
--
-- USDM SUB-UNITS: 1 USDM = 100 integer units.
-- pdPriceUsdm is stored in sub-units (e.g. Genesis = 100, Class 1 = 200).
-- The formula produces sub-unit results directly:
--   payout_subunits = base * priceUsdm / 2
--   payout_USDM = payout_subunits / 100
--
-- Relationship between ticket CLASS, TIER, and PAYOUT:
--
--   Ticket CLASS = pdPriceUsdm (100, 200, 300, 500, 1000, 2500, 5000, 10000 sub-units)
--   Winning TIER = classifyTier(symbols) (1..5, or 0 for loss)
--   PAYOUT (sub-units) = base * priceUsdm / 2
--
-- Example for Genesis (priceUsdm = 100 sub-units):
--   Tier 1: 2 * 100 / 2 = 100 sub-units = 1.00 USDM   (3+ of symbol 1)
--   Tier 2: 5 * 100 / 2 = 250 sub-units = 2.50 USDM  (3+ of symbol 2)
--   Tier 3: 10 * 100 / 2 = 500 sub-units = 5.00 USDM   (3+ of symbol 3)
--   Tier 4: 200 * 100 / 2 = 10000 sub-units = 100.00 USDM (3+ of symbol 4)
--   Tier 5: 1000 * 100 / 2 = 50000 sub-units = 500.00 USDM (3+ of symbol 5)
--
-- Example for Class 1 (priceUsdm = 200 sub-units):
--   Tier 1: 2 * 200 / 2 = 200 sub-units = 2.00 USDM
--   Tier 2: 5 * 200 / 2 = 500 sub-units = 5.00 USDM
--   etc.
--
-- This is NOT the same as "ticket class = tier". A 1 USDM ticket
-- can win tier 5 (500 USDM). A 100 USDM ticket can win tier 1 (100 USDM).
data PrizeTable = PrizeTable
  { ptTier1 :: Integer
  , ptTier2 :: Integer
  , ptTier3 :: Integer
  , ptTier4 :: Integer
  , ptTier5 :: Integer
  }

PlutusTx.unstableMakeIsData ''PrizeTable
PlutusTx.makeLift ''PrizeTable

{-# INLINABLE defaultPrizeTable #-}
defaultPrizeTable :: PrizeTable
defaultPrizeTable =
  PrizeTable
    { ptTier1 = 2
    , ptTier2 = 5
    , ptTier3 = 10
    , ptTier4 = 200
    , ptTier5 = 1000
    }

{-# INLINABLE baseForTier #-}
baseForTier :: PrizeTable -> Integer -> Integer
baseForTier t tier =
  if tier == 1 then ptTier1 t
  else if tier == 2 then ptTier2 t
  else if tier == 3 then ptTier3 t
  else if tier == 4 then ptTier4 t
  else if tier == 5 then ptTier5 t
  else 0

{-# INLINABLE prizeAmountForTier #-}
prizeAmountForTier :: PrizeTable -> Integer -> Integer -> Integer
prizeAmountForTier table tier priceUsdm =
  if tier <= 0 || priceUsdm <= 0
    then 0
    else (baseForTier table tier * priceUsdm) `divide` 2

-- ============================================================
-- Symbol generation
-- ============================================================

{-# INLINABLE countSym #-}
countSym :: BuiltinByteString -> Integer -> Integer -> Integer
countSym bs sym i =
  if i >= 6
    then 0
    else
      let c = indexByteString bs i
          n = if c == sym then 1 else 0
      in n + countSym bs sym (i + 1)

-- | Returns the highest symbol (5..1) appearing at least three times.
--   Returns 0 when there is no winning tier.
{-# INLINABLE classifyTier #-}
classifyTier :: BuiltinByteString -> Integer
classifyTier bs =
  if lengthOfByteString bs < 6
    then 0
    else go 5
  where
    go sym =
      if sym <= 0
        then 0
        else if countSym bs sym 0 >= 3
          then sym
          else go (sym - 1)

-- | Generates exactly six symbols in the range 1..5 using rejection sampling.
--
-- Each position hashes: sha2_256( byte(position) || symbolsSeed )
-- and consumes bytes sequentially from the 32-byte hash.
-- Bytes with value 255 are rejected (skipped) because 256 is not
-- evenly divisible by 5.
--
-- FAILS if more than 32 bytes are needed for a single position
-- (extremely unlikely: probability < (1/256)^32).
-- This bounds all hash reads within the SHA-256 output.
--
-- The result is completely deterministic from symbolsSeed.
{-# INLINABLE generateSymbols #-}
generateSymbols :: BuiltinByteString -> BuiltinByteString
generateSymbols symbolsSeed = collect 0 0 emptyByteString
  where
    nextByte pos hashPos =
      indexByteString
        (sha2_256 (appendByteString (consByteString pos emptyByteString) symbolsSeed))
        hashPos

    collect count _ acc
      | count >= 6 = acc
    collect count hashPos acc =
      if hashPos >= 32
        then traceError "generateSymbols: hash exhausted for position"
        else
          let byte = nextByte count hashPos
          in if byte == 255
               then collect count (hashPos + 1) acc
               else collect (count + 1) 0
                      (appendByteString acc (consByteString (remainder byte 5 + 1) emptyByteString))